bindings/sundials4py/idas/idas.cpp

```diff
@@ -0,0 +1,319 @@
+/*------------------------------------------------------------------------------
+ * Programmer(s): Cody J. Balos @ LLNL
+ *------------------------------------------------------------------------------
+ * SUNDIALS Copyright Start
+ * Copyright (c) 2025-2025, Lawrence Livermore National Security,
+ * University of Maryland Baltimore County, and the SUNDIALS contributors.
+ * Copyright (c) 2013-2025, Lawrence Livermore National Security
+ * and Southern Methodist University.
+ * Copyright (c) 2002-2013, Lawrence Livermore National Security.
+ * All rights reserved.
+ *
+ * See the top-level LICENSE and NOTICE files for details.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ * SUNDIALS Copyright End
+ *----------------------------------------------------------------------------*/
+
+#include "sundials4py.hpp"
+
+#include <sundials/sundials_core.hpp>
+
+#include <idas/idas.h>
+#include <idas/idas.hpp>
+#include <idas/idas_ls.h>
+
+#include "idas/idas_impl.h"
+
+#include "idas_usersupplied.hpp"
+
+namespace nb = nanobind;
+using namespace sundials::experimental;
+
+#define BIND_IDA_CALLBACK(NAME, FN_TYPE, MEMBER, WRAPPER, ...)          \
+  m.def(                                                                \
+    #NAME,                                                              \
+    [](void* ida_mem, std::function<std::remove_pointer_t<FN_TYPE>> fn) \
+    {                                                                   \
+      auto fn_table    = get_idas_fn_table(ida_mem);                    \
+      fn_table->MEMBER = nb::cast(fn);                                  \
+      if (fn) { return NAME(ida_mem, &WRAPPER); }                       \
+      else { return NAME(ida_mem, nullptr); }                           \
+    },                                                                  \
+    __VA_ARGS__)
+
+#define BIND_IDA_CALLBACK2(NAME, FN_TYPE1, MEMBER1, WRAPPER1, FN_TYPE2,   \
+                           MEMBER2, WRAPPER2, ...)                        \
+  m.def(                                                                  \
+    #NAME,                                                                \
+    [](void* ida_mem, std::function<std::remove_pointer_t<FN_TYPE1>> fn1, \
+       std::function<std::remove_pointer_t<FN_TYPE2>> fn2)                \
+    {                                                                     \
+      auto fn_table     = get_idas_fn_table(ida_mem);                     \
+      fn_table->MEMBER1 = nb::cast(fn1);                                  \
+      fn_table->MEMBER2 = nb::cast(fn2);                                  \
+      if (fn1) { return NAME(ida_mem, WRAPPER1, WRAPPER2); }              \
+      else { return NAME(ida_mem, nullptr, WRAPPER2); }                   \
```

Comment:
```suggestion
      if (fn1 && fn2) { return NAME(ida_mem, WRAPPER1, WRAPPER2); }       \
      else if (fn1) { return NAME(ida_mem, WRAPPER1, nullptr); }          \
      else if (fn2) { return NAME(ida_mem, nullptr, WRAPPER2); }          \
      else { return NAME(ida_mem, nullptr, nullptr); }                    \
```
and should these be `WRAPPER` or `&WRAPPER`?